File: 백준/Gold/14267. 회사 문화 1/회사 문화 1.java
import java.util.*;
import java.io.*;

class Main {

	public static void main(String[] args) throws Exception {
		final BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		StringTokenizer st = new StringTokenizer(br.readLine());

		final int n = Integer.parseInt(st.nextToken());
		final int m = Integer.parseInt(st.nextToken());

		st = new StringTokenizer(br.readLine());

		final ArrayList<Integer>[] graph = new ArrayList[n];
		final long[] comps = new long[n];

		for (int i = 0; i < n; i++) {
			final int upperNum = Integer.parseInt(st.nextToken());
			graph[i] = new ArrayList<>();
			comps[i] = 0;

			if (upperNum == -1) {
				continue;
			}

//			arr[upperNum - 1].lowers.add(emp);
			graph[upperNum - 1].add(i);
		}

		for (int i = 0; i < m; i++) {
			st = new StringTokenizer(br.readLine());

			final int num = Integer.parseInt(st.nextToken()) - 1;
			final long comp = Long.parseLong(st.nextToken());

			comps[num] += comp;
//			dfs(arr[num - 1], comp);
		}

		dfs(comps, graph, 0);

		for (final long comp : comps) System.out.print(comp + " ");
	}

	private static void dfs(long[] comps, ArrayList<Integer>[] graph, int num) {
		graph[num].forEach(child -> {
			comps[child] += comps[num];
			dfs(comps, graph, child);
		});
	}

}

File: 백준/Gold/14267. 회사 문화 1/회사 문화 1.kt
import java.io.*
import java.util.*
import kotlin.collections.ArrayList

/*
6 6
-1 1 2 3 3 2
2 123
3 123
4 123
5 123
2 123
3 123

답: 0 246 492 615 615 246

* */

fun main() {
    val br = BufferedReader(InputStreamReader(System.`in`))
    var st = StringTokenizer(br.readLine())

    val n = st.nextInt()
    val m = st.nextInt()

    st = StringTokenizer(br.readLine())

    val arr = LongArray(n)
    val graph = Array<ArrayList<Int>>(n) { ArrayList() }

    for (i in 0 until n) {
        val num = st.nextInt()
        if (num == -1) continue

//        arr[num - 1].lowerEmps.add(emp)
        graph[num - 1].add(i)
    }

    for (i in 0 until m) {
        st = StringTokenizer(br.readLine())

        val num = st.nextInt() - 1
        val com = st.nextInt().toLong()

        arr[num] += com
    }

    dfs(arr, graph, 0)

    arr.forEach(Long::printWithSeparate)
}

private fun Long.printWithSeparate() = print("$this ")

private fun dfs(arr: LongArray, graph: Array<ArrayList<Int>>, num: Int) {
    graph[num].forEach {
        arr[it] += arr[num]
        dfs(arr, graph, it)
    }
}

private fun BufferedReader.readInt(): Int = readLine().toInt()

private fun StringTokenizer.nextInt(): Int = nextToken().toInt()
